src/components/MainTable/MainTable.js

```diff
@@ -82,60 +84,64 @@ const generateRows = (
-  if (sortable && currentSortKey) {
-    if (!sortFunction) {
-      sortFunction = (a, b) => {
-        if (!a.sortData || !b.sortData) {
+  if (Object.entries(rows).length) {
+    const sortedRows = [...rows];
+    if (sortable && currentSortKey) {
+      if (!sortFunction) {
+        sortFunction = (a, b) => {
+          if (!a.sortData || !b.sortData) {
+            return 0;
+          }
+          if (a.sortData[currentSortKey] > b.sortData[currentSortKey]) {
+            return currentSortDirection === "ascending" ? 1 : -1;
+          } else if (a.sortData[currentSortKey] < b.sortData[currentSortKey]) {
+            return currentSortDirection === "ascending" ? -1 : 1;
+          }
           return 0;
-        }
-        if (a.sortData[currentSortKey] > b.sortData[currentSortKey]) {
-          return currentSortDirection === "ascending" ? 1 : -1;
-        } else if (a.sortData[currentSortKey] < b.sortData[currentSortKey]) {
-          return currentSortDirection === "ascending" ? -1 : 1;
-        }
-        return 0;
-      };
+        };
+      }
+      sortedRows.sort((a, b) =>
+        sortFunction(a, b, currentSortDirection, currentSortKey)
+      );
     }
-    sortedRows.sort((a, b) =>
-      sortFunction(a, b, currentSortDirection, currentSortKey)
-    );
-  }
-  let slicedRows = sortedRows;
-  if (paginate) {
-    let startIndex = (currentPage - 1) * paginate;
-    if (startIndex > rows.length) {
-      // If the rows have changed e.g. when filtering and the user is on a page
-      // that no longer exists then send them to the start.
-      setCurrentPage(1);
+    let slicedRows = sortedRows;
+    if (paginate) {
+      let startIndex = (currentPage - 1) * paginate;
+      if (startIndex > rows.length) {
+        // If the rows have changed e.g. when filtering and the user is on a page
+        // that no longer exists then send them to the start.
+        setCurrentPage(1);
+      }
+      slicedRows = sortedRows.slice(startIndex, startIndex + paginate);
     }
-    slicedRows = sortedRows.slice(startIndex, startIndex + paginate);
+    const rowItems = slicedRows.map(
+      (
+        { columns, expanded, expandedContent, key, sortData, ...rowProps },
+        index
+      ) => {
+        const cellItems = columns.map(({ content, ...cellProps }, index) => (
+          <TableCell key={index} {...cellProps}>
+            {content}
+          </TableCell>
+        ));
+        // The expanding cell is alway created to match the correct number of
+        // table cells in rows that do have expanding content.
+        return (
+          <TableRow key={key || index} {...rowProps}>
+            {cellItems}
+            {expanding && (
+              <TableCell expanding={true} hidden={!expanded}>
+                {expandedContent}
+              </TableCell>
+            )}
+          </TableRow>
+        );
+      }
+    );
+    return <tbody>{rowItems}</tbody>;
+  } else {
+    return <caption>No data to display</caption>;
```

Comment:
Why caption and not just text?

Reply:
Raw text dropped inside a `<table>` is semantically meaningless and AFAIK invalid. As per the [table HTML spec](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/table), I believe caption is the most appropriate choice here.

Reply:
I think this will need to be supplied from a prop.